README.md
## Was ist der Yellowcube?

Der Yellowcube ist eine automatisierte Warenhauslösung der Swiss Post AG. Sie ermöglicht die automatisierte Verarbeitung der Einlagerung, Lagerhaltung und den Versand von kleinen Gegenständen. Mehr Informationen zum Yellowcube finden Sie auf den Seiten der Swiss Post AG.

## A-SIGN Shopware2Yellowcube

Der Yellowcube Connector verbindet Shopware mit dem Warenhaussystem des Yellowcube. Durch die Integration wird es für einen Shop-Betreiber noch einfacher von den Vorteilen einer automatisierten Lagerlösung zu profitieren. Das vollständig lizensierte Yellowcube-Modul unterstützt alle Funktionen der Yellowcube-Schnittstelle auf der Shopware-Plattform, wird regelmässig gewartet und mit neuen Funktionen des Yellowcube erweitert.

## Funktionen

* Automatische Bestellübermittlung, sowie Statusabfrage an den Yellowcube
* Artikelstammanreicherung direkt über den Shop, sodass Ihre Waren bei der nächsten Anlieferung schon dem System bekannt sind. Kein Umweg über ein ERP notwendig
* Modernes SOAP-Interface mit Zertifikat für maximale Sicherheit
* Alle Yellowcube Versandarten und Optionen werden unterstützt
* Bestelldokumente werden direkt aus dem Shop als PDF an den Yellowcube übermittelt
* Automatische Lagerverwaltung und Synchronisierung der Lagerwerte.
* Versand von Tracking-Nr. E-Mails, damit Ihre Kunden den Weg der Ware verfolgen können
* Unterstützt Erstellung von PDF-Dokumenten für internationalen Versand (Handelsrechnung, CN22)

## Installation

Das Modul unterstützt die Shopware-Version 5.6.4. Kopieren sie den Ordner engine in ihr Shop Verzeichniss. 
Das Plugin wird danach im Plug-In Manager sichtbar für die Installation und Aktivierung sein.

Falls Ihr Shop eine spezielle Anpassung benötigt, stehen wir Ihnen sowohl bei der Installation, als auch bei der Anpassung des Moduls jederzeit zu Verfügung.

Kontaktieren sie uns unter: info@a-sign.ch
oder telefonisch unter +41 71 677 90 00

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## v2.3.0
- Compatible with SW 5.6.4
- Don't overwrite the shopware shop stock with Yellocube stock
- Added columns: Ycube Artnum, ArticleNr, EAN, Stock, YC Stock
- Sort by columns is possible
- Search by columns is possible
- Results page shows 50 results per page, if more are available then pagination is available too
- Transfer of YC stock to the shop
- Stock remains unchanged until carried out by Yellowcube
- Changes in WAB Name Mapping. Field Name4 doest exist anymore
- Removed Shopware_Components_Translation() and get the translation component from container: $this->container->get('translation')

## v2.2.2
- Don't export multilang article titles from inactive shops
- Don't export empty article titles from translations

## v2.2.1
- Insert article: fix multi translations

## v2.2.0
- Add infoline in orders
- Add article attribute-flag for YC export
- Add config for inventory reset
- Skip multiple inventory rows on update (performance/refactoring)

## v2.1.8
- Fix multiple shop bug

## v2.1.7
- Remove backend config "use certificate"
- Fix some code definitions for PHP 7 support

## v2.1.6
- Fix track&trace numbers for orders

## v2.1.5
- Ignore/skip vouchers when transferring to YellowCube
- Set last order state to "complete" if WAR response is success

## v2.1.4
- Validate max field length in order address data
- Run Inventory cronjob only once a day
- Rename Cronjobs

## v2.1.3
- Fix article bulk send & response message in backend

## v2.1.2
- Bugfix backend order buttons and actions

## v2.1.1
- Cleanup and simplify code



engine/Shopware/Plugins/Local/Backend/AsignYellowcube/Models/AsignModels/Orders/Orders.php
<?php

/**
 * This file defines data model for Orders
 *
 * PHP version 5
 *
 * @category  asign
 * @package   AsignYellowcube
 * @author    entwicklung@a-sign.ch
 * @copyright A-Sign
 * @license   https://www.a-sign.ch/
 * @version   2.1
 * @link      https://www.a-sign.ch/
 * @see       Orders
 * @since     File available since Release 1.0
 */

namespace Shopware\CustomModels\AsignModels\Orders;

use Doctrine\ORM\Mapping as ORM;
use Shopware\Components\Model\ModelEntity;

/**
 * Defines data model for Orders
 *
 * @category A-Sign
 * @package  AsignYellowcube
 * @author   entwicklung@a-sign.ch
 * @link     http://www.a-sign.ch
 */

/**
 * Shopware\CustomModels\AsignModels\Orders\Orders
 *
 * @ORM\Table(name="asign_yellowcube_orders")
 * @ORM\Entity(repositoryClass="Repository")
 */
class Orders extends ModelEntity {
	/**
	 * @var integer $id
	 *
	 * @ORM\Column(name="id", type="integer", precision=0, scale=0, nullable=false, unique=false)
	 * @ORM\Id
	 * @ORM\GeneratedValue(strategy="IDENTITY")
	 */
	private $id;

	/**
	 * @var integer $ordid
	 *
	 * @ORM\Column(name="ordid", type="integer", length=10, precision=0, scale=0, nullable=false, unique=false)
	 */
	private $ordid;

	/**
	 * @var integer $lastSent
	 *
	 * @ORM\Column(name="lastSent", type="boolean", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $lastSent = false;

	/**
	 * @var string $ycReference
	 *
	 * @ORM\Column(name="ycReference", type="text", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $ycReference;

	/**
	 * @var string $ycResponse
	 *
	 * @ORM\Column(name="ycResponse", type="text", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $ycResponse = null;

	/**
	 * @var string $ycWabResponse
	 *
	 * @ORM\Column(name="ycWabResponse", type="text", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $ycWabResponse = null;

	/**
	 * @var string $ycWarResponse
	 *
	 * @ORM\Column(name="ycWarResponse", type="text", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $ycWarResponse = null;

	/**
	 * @var \DateTime $createDate
	 *
	 * @ORM\Column(name="createDate", type="datetime", precision=0, scale=0, nullable=false, unique=false))
	 */
	private $createDate = null;

	/**
	 * @return int
	 */
	public function getId() {
		return $this->id;
	}

	/**
	 * @return int
	 */
	public function getOrdid() {
		return $this->ordid;
	}

	/**
	 * @param int $ordid
	 */
	public function setOrdid( $ordid ) {
		$this->ordid = $ordid;
	}

	/**
	 * @return int
	 */
	public function getLastSent() {
		return $this->lastSent;
	}

	/**
	 * @param int $lastSent
	 */
	public function setLastSent( $lastSent ) {
		$this->lastSent = $lastSent;
	}

	/**
	 * @return string
	 */
	public function getYcReference() {
		return $this->ycReference;
	}

	/**
	 * @param string $ycReference
	 */
	public function setYcReference( $ycReference ) {
		$this->ycReference = $ycReference;
	}

	/**
	 * @return string
	 */
	public function getYcResponse() {
		return $this->ycResponse;
	}

	/**
	 * @param string $ycresponse
	 */
	public function setYcResponse( $ycresponse ) {
		$this->ycResponse = $ycresponse;
	}

	/**
	 * @return string
	 */
	public function getYcWabResponse() {
		return $this->ycWabResponse;
	}

	/**
	 * @param string $ycwabresponse
	 */
	public function setYcWabResponse( $ycwabresponse ) {
		$this->ycWabResponse = $ycwabresponse;
	}

	/**
	 * @return string
	 */
	public function getYcWarResponse() {
		return $this->ycWarResponse;
	}

	/**
	 * @param string $ycwarresponse
	 */
	public function setYcWarResponse( $ycwarresponse ) {
		$this->ycWarResponse = $ycwarresponse;
	}

	/**
	 * @return \DateTime
	 */
	public function getCreateDate() {
		return $this->createDate;
	}

	/**
	 * @param \DateTime $createDate
	 */
	public function setCreateDate( $createDate ) {
		$this->createDate = $createDate;
	}
}


engine/Shopware/Plugins/Local/Backend/AsignYellowcube/Models/AsignModels/Product/Product.php
<?php

/**
 * This file defines data model for Products
 *
 * PHP version 5
 *
 * @category  asign
 * @package   AsignYellowcube
 * @author    entwicklung@a-sign.ch
 * @copyright A-Sign
 * @license   https://www.a-sign.ch/
 * @version   2.1
 * @link      https://www.a-sign.ch/
 * @see       Product
 * @since     File available since Release 1.0
 */

namespace Shopware\CustomModels\AsignModels\Product;

use Doctrine\ORM\Mapping as ORM;
use Shopware\Components\Model\ModelEntity;

/**
 * Defines data model for Products
 *
 * @category A-Sign
 * @package  AsignYellowcube
 * @author   entwicklung@a-sign.ch
 * @link     http://www.a-sign.ch
 */

/**
 * Shopware\CustomModels\AsignModels\Product\Product
 *
 * @ORM\Table(name="asign_yellowcube_product")
 * @ORM\Entity(repositoryClass="Repository")
 */
class Product extends ModelEntity {
	/**
	 * @var integer $id
	 *
	 * @ORM\Column(name="id", type="integer", precision=0, scale=0, nullable=false, unique=false)
	 * @ORM\Id
	 * @ORM\GeneratedValue(strategy="IDENTITY")
	 */
	private $id;

	/**
	 * @var integer $artid
	 *
	 * @ORM\Column(name="artid", type="integer", length=10, precision=0, scale=0, nullable=false, unique=false)
	 */
	private $artid;

	/**
	 * @var integer $lastSent
	 *
	 * @ORM\Column(name="lastSent", type="boolean", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $lastSent = false;

	/**
	 * @var string $ycSpsDetails
	 *
	 * @ORM\Column(name="ycSpsDetails", type="text", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $ycSpsDetails = null;

	/**
	 * @var string $ycResponse
	 *
	 * @ORM\Column(name="ycResponse", type="text", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $ycResponse = null;

	/**
	 * @var string $ycReference
	 *
	 * @ORM\Column(name="ycReference", type="integer", length=10, precision=0, scale=0, nullable=false, unique=false)
	 */
	private $ycReference;

	/**
	 * @var \DateTime $createDate
	 *
	 * @ORM\Column(name="createDate", type="datetime", precision=0, scale=0, nullable=false, unique=false))
	 */
	private $createDate = null;

	/**
	 * @return int
	 */
	public function getId() {
		return $this->id;
	}

	/**
	 * @return int
	 */
	public function getArtid() {
		return $this->artid;
	}

	/**
	 * @param int $artid
	 */
	public function setArtid( $artid ) {
		$this->artid = $artid;
	}

	/**
	 * @return int
	 */
	public function getLastSent() {
		return $this->lastSent;
	}

	/**
	 * @param int $lastSent
	 */
	public function setLastSent( $lastSent ) {
		$this->lastSent = $lastSent;
	}

	/**
	 * @return string
	 */
	public function getYcSpsDetails() {
		return $this->ycSpsDetails;
	}

	/**
	 * @param string $ycSpsDetails
	 */
	public function setYcSpsDetails( $ycSpsDetails ) {
		$this->ycSpsDetails = $ycSpsDetails;
	}

	/**
	 * @return string
	 */
	public function getYcResponse() {
		return $this->ycResponse;
	}

	/**
	 * @param string $ycresponse
	 */
	public function setYcResponse( $ycresponse ) {
		$this->ycResponse = $ycresponse;
	}

	/**
	 * @return string
	 */
	public function getYcReference() {
		return $this->ycReference;
	}

	/**
	 * @param string $ycReference
	 */
	public function setYcReference( $ycReference ) {
		$this->ycReference = $ycReference;
	}

	/**
	 * @return \DateTime
	 */
	public function getCreateDate() {
		return $this->createDate;
	}

	/**
	 * @param \DateTime $createDate
	 */
	public function setCreateDate( $createDate ) {
		$this->createDate = $createDate;
	}
}


engine/Shopware/Plugins/Local/Backend/AsignYellowcube/Models/AsignModels/Inventory/Inventory.php
<?php

/**
 * This file defines data model for Inventory
 *
 * PHP version 5
 *
 * @category  asign
 * @package   AsignYellowcube
 * @author    entwicklung@a-sign.ch
 * @copyright A-Sign
 * @license   https://www.a-sign.ch/
 * @version   2.1
 * @link      https://www.a-sign.ch/
 * @see       Inventory
 * @since     File available since Release 1.0
 */

namespace Shopware\CustomModels\AsignModels\Inventory;

use Doctrine\ORM\Mapping as ORM;
use Shopware\Components\Model\ModelEntity;

/**
 * Defines data model for Inventory
 *
 * @category A-Sign
 * @package  AsignYellowcube
 * @author   entwicklung@a-sign.ch
 * @link     http://www.a-sign.ch
 */

/**
 * Shopware\CustomModels\AsignModels\Inventory\Inventory
 *
 * @ORM\Table(name="asign_yellowcube_inventory")
 * @ORM\Entity(repositoryClass="Repository")
 */
class Inventory extends ModelEntity {
	/**
	 * @var integer $id
	 *
	 * @ORM\Column(name="id", type="integer", precision=0, scale=0, nullable=false, unique=false)
	 * @ORM\Id
	 * @ORM\GeneratedValue(strategy="IDENTITY")
	 */
	private $id;

	/**
	 * @var string $ycarticlenr
	 *
	 * @ORM\Column(name="ycarticlenr", type="string", length=128, precision=0, scale=0, nullable=false, unique=false)
	 */
	private $ycarticlenr;

	/**
	 * @var string $articlenr
	 *
	 * @ORM\Column(name="articlenr", type="string", length=100, precision=0, scale=0, nullable=false, unique=false)
	 */
	private $articlenr;

	/**
	 * @var string $artdesc
	 *
	 * @ORM\Column(name="artdesc", type="text", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $artdesc;

	/**
	 * @var string $additional
	 *
	 * @ORM\Column(name="additional", type="text", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $additional;

	/**
	 * @var \DateTime $createdon
	 *
	 * @ORM\Column(name="createdon", type="datetime", precision=0, scale=0, nullable=false, unique=false)
	 */
	private $createdon;

	/**
	 * Get id
	 *
	 * @return integer
	 */
	public function getId() {
		return $this->id;
	}

	/**
	 * Set ycarticlenr
	 *
	 * @param string $sValue
	 *
	 * @return Inventory
	 */
	public function setYcartnum( $sValue ) {
		if ( ! empty( $sValue ) ) {
			$this->ycarticlenr = $sValue;
		}

		return $this;
	}

	/**
	 * Get Ycarticlenr
	 *
	 * @return string
	 */
	public function getYcartnum() {
		return $this->ycarticlenr;
	}

	/**
	 * Set articlenr
	 *
	 * @param string $sValue
	 *
	 * @return Inventory
	 */
	public function setArtnum( $sValue ) {
		if ( ! empty( $sValue ) ) {
			$this->articlenr = $sValue;
		}

		return $this;
	}

	/**
	 * Get Article Number
	 *
	 * @return string
	 */
	public function getArtnum() {
		return $this->articlenr;
	}

	/**
	 * Get Article Description
	 *
	 * @return string
	 */
	public function getArtDesc() {
		return $this->artdesc;
	}

	/**
	 * Set article desc
	 *
	 * @param string $sValue
	 *
	 * @return Inventory
	 */
	public function setArtDesc( $sValue ) {
		if ( ! empty( $sValue ) ) {
			$this->artdesc = $sValue;
		}

		return $this;
	}

	/**
	 * Set additional information
	 *
	 * @param string $sValue
	 *
	 * @return Inventory
	 */
	public function setAddInfo( $sValue ) {
		if ( ! empty( $sValue ) ) {
			$this->additional = $sValue;
		}

		return $this;
	}

	/**
	 * Get Additional Information
	 *
	 * @return text
	 */
	public function getAddInfo() {
		return $this->additional;
	}

	/**
	 * Set created
	 *
	 * @param \DateTime $created
	 *
	 * @return Inventory
	 */
	public function setCreated( $created ) {
		if ( ! empty( $created ) ) {
			$this->createdon = $created;
		}

		return $this;
	}

	/**
	 * Get created
	 *
	 * @return \DateTime
	 */
	public function getCreated() {
		return $this->createdon;
	}
}


engine/Shopware/Plugins/Local/Backend/AsignYellowcube/Models/AsignModels/Inventory/Repository.php
<?php
/**
 * This file defines data repository for Inventory
 *
 * @category  asign
 * @package   AsignYellowcube
 * @author    entwicklung@a-sign.ch
 * @copyright A-Sign
 * @license   https://www.a-sign.ch/
 * @version   2.1.3
 * @link      https://www.a-sign.ch/
 * @since     File available since Release 1.0
 */

namespace Shopware\CustomModels\AsignModels\Inventory;

use Shopware\AsignYellowcube\Helpers\ApiClasses\AsignSoapClientApi;
use Shopware\Components\Model\ModelRepository;

/**
 * Defines repository for Inventory
 *
 * @category A-Sign
 * @package  AsignYellowcube
 * @author   entwicklung@a-sign.ch
 * @link     http://www.a-sign.ch
 */
class Repository extends ModelRepository {

	/**
	 * Method to query products and return result.
	 *
	 * @param $filters
	 * @param $sort
	 * @param int $offset
	 * @param int $limit
	 *
	 * @return array
	 */
	public function getInventoryListData( $filters, $sort, $offset = 0, $limit = 100 ) {
		$select = $this->getInventoryListQuery( $filters, $sort, $offset, $limit );
		$result = [];

		// set the paginator and result
		$paginator       = new \Zend_Paginator_Adapter_DbSelect( $select );
		$data            = $select->query()->fetchAll();
		$result['data']  = $data;
		$result['total'] = $paginator->count();

		return $result;
	}


	/**
	 * Returns all the inventory based on filter or sort.
	 *
	 * @param array $filters Filters
	 * @param integer $sort Sort value
	 * @param integer $offset Offset value
	 * @param integer $limit Limit value
	 *
	 * @return array
	 */
	public function getInventoryListQuery( $filters, $sort, $offset = 0, $limit = 100 ) {
		$select = Shopware()->Db()->select()
		                    ->from( 'asign_yellowcube_inventory' )->limit( $limit, $offset );

		//If a filter is set
		if ( $filters ) {
			foreach ( $filters as $filter ) {
				$select->where( 'asign_yellowcube_inventory.ycarticlenr LIKE ?', '%' . $filter["value"] . '%' );
				$select->orWhere( 'asign_yellowcube_inventory.articlenr LIKE ?', '%' . $filter["value"] . '%' );
				$select->orWhere( 'asign_yellowcube_inventory.artdesc LIKE ?', '%' . $filter["value"] . '%' );
				$select->orWhere( 'asign_yellowcube_inventory.additional LIKE ?', '%' . $filter["value"] . '%' );
			}
		}

		// sortin the inventory list
		if ( $sort ) {
			$sorting = reset( $sort );
			switch ( $sorting['property'] ) {
				case 'ycarticlenr':
					$select->order( 'asign_yellowcube_inventory.ycarticlenr ' . $sorting['direction'] );
					break;
				case 'articlenr':
					$select->order( 'asign_yellowcube_inventory.articlenr ' . $sorting['direction'] );
					break;
				case 'artdesc':
					$select->order( 'asign_yellowcube_inventory.artdesc ' . $sorting['direction'] );
					break;
				default:
					$select->order( 'asign_yellowcube_inventory.createdon ' . $sorting['direction'] );
			}
		} else {
			$select->order( 'asign_yellowcube_inventory.createdon DESC' );
		}

		return $select;
	}

	/**
	 * Stores inventory information received from Yellowcube
	 *
	 * @param array $aResponseData Object of response
	 *
	 * @return integer Updated articles
	 */
	public function saveInventoryData( $aResponseData ) {
		// format the response data
		$iCount = 0;

		// reset the inventory data
		$oSoapApi = new AsignSoapClientApi();
		if ( $oSoapApi->getYCResetInventory() ) {
			$this->resetInventoryData();
		}

		foreach ( $aResponseData['ArticleList']['Article'] as $aArticle ) {
			//skip multiple article rows, only location YAFS and type blank is needed
			if ( ! empty( $aArticle['ArticleNo'] && $aArticle['StorageLocation'] == 'YAFS' && ($aArticle['StockType'] == '0' || $aArticle['StockType'] == 'F' || $aArticle['StockType'] == '') ) ) {
				$artnr = $aArticle['ArticleNo'];

				$sGetArtid = "SELECT s_articles_details.articleID FROM s_articles_details LEFT JOIN s_articles_attributes ON s_articles_details.articleID = s_articles_attributes.articleID AND s_articles_attributes.yc_export = 1 WHERE s_articles_details.ordernumber = ?";
				$artid     = Shopware()->Db()->executeQuery( $sGetArtid, [ $artnr ] )->fetchColumn();

				// skip if no SW article is found (check yc_export = 1 in article attributes)
				if ( empty( $artid ) ) {
					continue;
				}

				$qtyISO  = $aArticle['QuantityUOM']['QuantityISO'];
				$qtyUOM  = $aArticle['QuantityUOM']['_'];
				$ycartnr = $aArticle['YCArticleNo'];

				$artdesc = $aArticle['ArticleDescription'];

				// entry id to avoid duplicates
				$mainId = substr( $ycartnr, 4 );

				// frame the additioanal information array
				$aAddInfo = [
					'EAN'             => $aArticle['EAN'],
					'Plant'           => $aArticle['Plant'],
					'StorageLocation' => $aArticle['StorageLocation'],
					'StockType'       => $aArticle['StockType'],
					'QuantityISO'     => $qtyISO,
					'QuantityUOM'     => $qtyUOM,
					'YCLot'           => $aArticle['YCLot'],
					'Lot'             => $aArticle['Lot'],
					'BestBeforeDate'  => $aArticle['BestBeforeDate'],
				];
				// serialize the data
				$sAdditional = serialize( $aAddInfo );

				// push in db
				$sUpdateYCInventory = "INSERT INTO `asign_yellowcube_inventory` SET `id` = '" . $mainId . "', `ycarticlenr` = '" . $ycartnr . "', `articlenr` = '" . $artnr . "', `artdesc` = '" . $artdesc . "', `additional` = '" . $sAdditional . "' ON DUPLICATE KEY UPDATE `createdon` = NOW(), `additional` = '".$sAdditional."'";
				Shopware()->Db()->query( $sUpdateYCInventory );

				//update the stock
				$iStock              = (int) $qtyUOM;
				$sUpdateArtInventory = "UPDATE `s_articles_details` SET `instock` = '" . $iStock . "' WHERE `ordernumber` = '" . $artnr . "'";
				Shopware()->Db()->query( $sUpdateArtInventory );

				$iCount ++;
			}
		}

		return $iCount;
	}

	/**
	 * Resets the oxstock value for all articles that are entered in the YC warehouse.
	 * This should be run before setting stock, because YC only sends information on articles, that have
	 * over 0 stock.
	 */
	public function resetInventoryData() {
		$aArticles = Shopware()->Db()->fetchAll( "SELECT `artid`, `ycResponse` FROM `asign_yellowcube_product` WHERE `ycResponse` != ''" );

		foreach ( $aArticles as $article ) {
			$aResponse = unserialize( $article['ycResponse'] );

			// convert object 2 array
			if ( is_object( $aResponse ) ) {
				$aResponse = json_decode( json_encode( $aResponse ), true );
			}

			if ( $aResponse['StatusCode'] == 100 ) {
				Shopware()->Db()->query( "UPDATE `s_articles_details` SET `instock` = '0' WHERE `articleID` = '" . $article['artid'] . "'" );
			}
		}
	}
}


engine/Shopware/Plugins/Local/Backend/AsignYellowcube/Components/PdfaDocument.php
<?php
/**
 * This file extends Shopware_Components_Document class
 *
 * PHP version 5
 *
 * @category  asign
 * @package   AsignYellowcube
 * @author    entwicklung@a-sign.ch
 * @copyright asign
 * @license   https://www.a-sign.ch/
 * @version   2.1
 * @link      https://www.a-sign.ch/
 * @see       PdfaDocument
 * @since     File available since Release 1.0
 */

require_once( Shopware()->AppPath( "Components" ) . "/Document.php" );
require_once( Shopware()->AppPath( "Components" ) . "/Translation.php" );

/**
 * Extends Shopware_Components_Document class
 *
 * @category Asign
 * @package  AsignYellowcube
 * @author   entwicklung@a-sign.ch
 * @link     http://www.a-sign.ch
 */
class PdfaDocument extends Shopware_Components_Document {
	/**
	 * Object from Type Model\Order
	 *
	 * @var object Model\Order
	 */
	public $_order;

	/**
	 * Shopware Template Object (Smarty)
	 *
	 * @var object
	 */
	public $_template;

	/**
	 * Shopware View Object (Smarty)
	 *
	 * @var object
	 */
	public $_view;

	/**
	 * Configuration
	 * @var array
	 */
	public $_config;

	/**
	 * compatibilityMode = true means that html2ps will be used instead of mpdf.
	 * Additionally old templatebase will be used (For pre 3.5 versions)
	 *
	 * Unsupported till shopware 4.0.0
	 * @var bool
	 * @deprecated
	 */
	public $_compatibilityMode = false;

	/**
	 * Define output
	 *
	 * @var string html,pdf,return
	 */
	public $_renderer = "html";

	/**
	 * Are properties already assigned to smarty?
	 *
	 * @var bool
	 */
	public $_valuesAssigend = false;

	/**
	 * Subshop-Configuration
	 *
	 * @var array
	 */
	public $_subshop;

	/**
	 * Path to load templates from
	 *
	 * @var string
	 */
	public $_defaultPath = "templates/_emotion";

	/**
	 * Generate preview only
	 *
	 * @var bool
	 */
	public $_preview = false;

	/**
	 * Typ/ID of document [0,1,2,3] - s_core_documents
	 *
	 * @var int
	 */
	public $_typID = 5;

	/**
	 * Document-Metadata / Properties
	 *
	 * @var array
	 */
	public $_document;

	/**
	 * Invoice / Document number
	 *
	 * @var int
	 */
	public $_documentID;

	/**
	 * Primary key of the created document row (s_order_documents)
	 *
	 * @var int
	 */
	public $_documentRowID;

	/**
	 * Hash of the created document row (s_order_documents.hash), will be used as filename when preview is false
	 *
	 * @var string
	 */
	public $_documentHash;

	/**
	 * Invoice ID for reference in shipping documents etc.
	 *
	 * @var string
	 */
	public $_documentBid;

	/**
	 * Ref to the translation component
	 *
	 * @var \Shopware_Components_Translation
	 */
	public $translationComponent;

	/**
	 * Start renderer / pdf-generation especially for PDFA
	 *
	 * @param string optional define renderer mostly an array
	 * with orderid and renderer
	 */
	public function pdfaRender( $aRenderer = null ) {
		// array of render and order-id data
		try {
			if ( $aRenderer ) {
				$_renderer      = $aRenderer['render'];
				$orderID        = $aRenderer['orderid'];
				$this->_preview = $aRenderer['preview'];
			}

			// set config details in session
			// get active Shop details
			$repository  = Shopware()->Models()->getRepository( 'Shopware\Models\Shop\Shop' );
			$defaultShop = $repository->getDefault();
			$localISO    = $defaultShop->getLocale()->getLocale();
			$shopID      = $defaultShop->getId();
			Shopware()->Models()->clear();

			// get Shop owner details: below ids are related to s_core_config_elements
			// and the values are stored in s_core_config_values..
			// 673 - shopName
			// 674 - mail
			// 675 - address
			// 676 - taxNumber
			// 677 - bankAccount
			// 893 - company
			$aShopData['shopname']          = $this->getConfigValue( 'shopName', $shopID );
			$aShopData['company']           = $this->getConfigValue( 'company', $shopID );
			$aShopData['mail']              = $this->getConfigValue( 'mail', $shopID );
			$aShopData['address']           = $this->getConfigValue( 'address', $shopID );
			$aShopData['taxnum']            = $this->getConfigValue( 'taxNumber', $shopID );
			$aShopData['bankaccnt']         = $this->getConfigValue( 'bankAccount', $shopID );
			$aShopData['country']           = $defaultShop->getLocale()->getTerritory();
			Shopware()->Session()->shopData = $aShopData;

			// get and set the countryId related to localeID
			$sCountISO                          = end( explode( "_", $localISO ) );
			$localId                            = Shopware()->Db()->fetchOne( "SELECT `id` FROM `s_core_countries` WHERE `countryiso` = '" . $sCountISO . "'" );
			Shopware()->Session()->shopLocaleId = (int) $localId;

			// pluginpath assigned
			$sShopPath = $defaultShop->getHost();
			if ( $defaultShop->getBasePath() ) {
				$sShopPath .= '/' . $defaultShop->getBasePath();
			}

			$pluginPath                       = 'http://' . $sShopPath . '/engine/Shopware/Plugins/Local/Backend/AsignYellowcube/Media/';
			Shopware()->Session()->pluginPath = $pluginPath;

			// start PDF-generation process
			$this->_renderer = $_renderer;
			$mpdfPath        = Shopware()->AppPath( 'Plugins' ) . "Local/Backend/AsignYellowcube/Helpers/Library/AsignMpdf/mpdf.php";
			include_once( $mpdfPath );

			//$this->_view = new Enlight_Template_Manager();
			$this->initTemplateEngine();

			// set the translation component
			$this->translationComponent = new Shopware_Components_Translation();
			$this->setOrder( Enlight_Class::Instance( 'Shopware_Models_Document_Order', [ $orderID, $config ] ) );

			if ( ! empty( $_renderer ) ) {
				$this->_renderer = $_renderer;
			}
			if ( $this->_valuesAssigend == false ) {
				$this->assignValues();
			}

			$sPdfTemplate = "engine/Shopware/Plugins/Local/Backend/AsignYellowcube/Views/";
			$data         = $this->_template->fetch( $sPdfTemplate . "documents/index_yc.tpl", $this->_view );
			if ( $this->_renderer == "html" || ! $this->_renderer ) {
				echo $data;
			} elseif ( $this->_renderer == "pdf" ) {
				if ( $this->_preview == true || ! $this->_documentHash ) {
					$mpdf = new AsignMpdf( "utf-8", "A4", "", "helvetica", $this->_document["left"], $this->_document["right"], $this->_document["top"], $this->_document["bottom"] );
					$mpdf->WriteHTML( $data );
					$mpdf->Output();
					die;
				} else {
					$path = Shopware()->OldPath() . "files/documents" . "/" . $this->_documentHash . ".pdf";
					$mpdf = new AsignMpdf( "utf-8", "A4", "", "helvetica", $this->_document["left"], $this->_document["right"], $this->_document["top"], $this->_document["bottom"] );
					$mpdf->WriteHTML( $data );
					$mpdf->Output( $path, "F" );

					// update order_documents table for correct bill-number
					$this->updateOrderBillNumber( $orderID, $this->_documentHash );
				}
			}
		} catch ( \Exception $ex ) {
			$oLogs = Shopware()->Models()->getRepository( "Shopware\CustomModels\AsignModels\Errorlogs\Errorlogs" );
			$oLogs->saveLogsData( 'pdfaRender', $ex );
		}
	}

	/**
	 * Return config value based on the name passed on and shopid
	 *
	 * @param string $sElement Form Element name
	 * @param integer $iShopid Shop ID
	 */
	protected function getConfigValue( $sElement, $iShopid ) {
		$sValue = Shopware()->Db()->fetchOne( "SELECT `value` FROM `s_core_config_values` WHERE `element_id` = (SELECT `id` FROM `s_core_config_elements` WHERE `name` = '" . $sElement . "') AND `shop_id` = '" . $iShopid . "'" );
		$sValue = unserialize( $sValue );

		return $sValue;
	}

	/**
	 * Update order billing number based on the orderid and hashid
	 * since somehow the billing number is not updated
	 *
	 * @param int $orderID order id
	 * @param string $sHashID File hash-id
	 */
	protected function updateOrderBillNumber( $orderID, $sHashID ) {
		// get the max current value in the list...
		$iMaxId = Shopware()->Db()->fetchOne( "SELECT MAX(docID) FROM `s_order_documents`" );
		$iMaxId = $iMaxId + 1;

		// update generated document details
		Shopware()->Db()->query( "UPDATE `s_order_documents` SET `docID` = '" . $iMaxId . "' WHERE `orderID` = '" . $orderID . "' AND `hash` = '" . $sHashID . "' AND `docID` = 0" );
	}
}
